### PR TITLE
Signed-off-by: Herbert Hum <hhum@sc-xterm-46.nvidia.com>

### PR DESCRIPTION
**Pull request by herberthum**
Fixed blas.py to support differing source types.
Fixed linear.py to support tensors of shape [1, X] as well as [X].

### CODE REVIEW

File: pyprof/prof/blas.py [hunk 1]
```diff
@@ -132,7 +132,11 @@ def __init__(self, d):
         assert (k1 == k2)
         t1 = A['dtype']
         t2 = B['dtype']
-        assert (t1 == t2)
+        # assert (t1 == t2).  No longer true.
+        if t1 != t2:
+            if t1 == 'float16' or t2 == 'float16' :
+                A['dtype'] = 'float16'
+                B['dtype'] = 'float16'
```
Comment (adityaiitb) [inline pyprof/prof/blas.py:139]: Can you provide a simple test case?

Comment (adityaiitb) [reply pyprof/prof/blas.py:139]: If this is true, it will impact all other matmul operations as well, not just bmm.

File: pyprof/prof/linear.py [hunk 2]
```diff
@@ -50,7 +50,10 @@ def setXWBMNK(self, args):
             x, w, b = args
             assert (x['type'] == w['type'] == "tensor")
             if (b['type'] == "tensor"):
-                assert (len(b['shape']) == 1)
+                # no longer true as b can be of shape [1, XX]
+                # assert (len(b['shape']) == 1)
+                tmp = 1 # dummy statement for commented line above
+                        # TODO: clean up if..elif..else struct
```
Comment (adityaiitb) [inline pyprof/prof/linear.py:56]: Please provide complete code or use the Python keyword `pass`.
Can you also provide a test case.

File: pyprof/prof/linear.py [hunk 3]
```diff
@@ -64,10 +67,13 @@ def setXWBMNK(self, args):
...
         if b is not None:
-            assert (b['shape'][0] == n)
+            # assert (b['shape'][0] == n)
+            assert(b['shape'][0] == n or 
+                   (b['shape'][0] == 1 and b['shape'][1] == n))
         t1 = x['dtype']
         t2 = w['dtype']
-        assert (t1 == t2)
+        # no longer true
+        # assert (t1 == t2)
```
Comment (adityaiitb) [inline pyprof/prof/linear.py:76]: Please provide a test case.

File: pyprof/prof/linear.py [hunk 4]
```diff
@@ -64,10 +67,13 @@ def setXWBMNK(self, args):
         n, k2 = w['shape']
         assert (k1 == k2)
         if b is not None:
-            assert (b['shape'][0] == n)
+            # assert (b['shape'][0] == n)
+            assert(b['shape'][0] == n or 
+                   (b['shape'][0] == 1 and b['shape'][1] == n))
```
Comment (adityaiitb) [inline pyprof/prof/linear.py:72]: `b['shape'] == [n] or b['shape'] == [1,n]` would be cleaner, I guess.

**Comment by adityaiitb** [comment]
Did the dtype assertion trigger because Ampere + Pytorch allows that? Are different dtypes allowed on Volta or Turing?

**Comment by herberthum** [comment]
Aditya,
I'm not at liberty to share our internal example where pyprof fails, but
those bugs are real.

Sincerely,
Herbert

File: pyprof/prof/blas.py (hunk 1 above)
Comment (dzier) [reply pyprof/prof/blas.py:139]: Also, can you provide a link to documentation that indicates this is not true? The comment is both obvious (i.e. "assert (t1 == t2)" and the code `t1 != t2`, the comment provide no additional insight). But "No longer true" is vague. Why is this no longer true?

File: pyprof/prof/linear.py (hunk 3 above)
Comment (dzier) [reply pyprof/prof/linear.py:76]: Do not leave commented code. Please delete if no longer needed.

**Review by dzier — CHANGES_REQUESTED** [review]
Please add unit tests for each of these new cases.

**Comment by adityaiitb** [comment]
@herberthum Please do not share anything confidential. I am asking for a 2-3 line code snippet (with random tensor dimensions) which triggers this assertion. These assertions have never failed in over 2 years and so I am very curious.

**Comment by adityaiitb** [comment]
My guess is that you are using AMP and PyProf and did not initialize them in the correct sequence. See the [FAQ](https://github.com/NVIDIA/PyProf/blob/master/docs/faqs.rst).

**Comment by herberthum** [comment]
Thanks for the pointer.  I need to figure out when all the JIT is done.

**Comment by adityaiitb** [comment]
Is this still an issue?

**Comment by herberthum** [comment]
We have an internal fix.  Thanks for checking.

**Pull request closed without merging**